diff --git a/vllm/entrypoints/openai/engine/protocol.py b/vllm/entrypoints/openai/engine/protocol.py
index 6b5b714dc32aa67e139546f8f24ad80cc8b6fe47..86300686b495cb53d1c29365b529b7a52b40d37c 100644
--- a/vllm/entrypoints/openai/engine/protocol.py
+++ b/vllm/entrypoints/openai/engine/protocol.py
@@ -158,7 +158,7 @@ AnyResponseFormat: TypeAlias = (
 
 
 class StreamOptions(OpenAIBaseModel):
-    include_usage: bool | None = True
+    include_usage: bool | None = False
     continuous_usage_stats: bool | None = False
 
 
