diff --git a/vllm/model_executor/layers/utils.py b/vllm/model_executor/layers/utils.py
index d1e35f5830c320f1fe67d19ecf695bd1b6ca61c0..d4357a00f0e8caa4bf81ea05b1bb56499329784a 100644
--- a/vllm/model_executor/layers/utils.py
+++ b/vllm/model_executor/layers/utils.py
@@ -210,7 +210,7 @@ direct_register_custom_op(
 
 def check_cpu_sgl_kernel(n: int, k: int, dtype: torch.dtype) -> bool:
     return (
-        torch._C._cpu._is_amx_tile_supported()
+        torch.cpu._is_amx_tile_supported()
         and (dtype in (torch.bfloat16, torch.int8))
         and k % 32 == 0
         and n % 16 == 0
