diff --git a/vllm/v1/engine/coordinator.py b/vllm/v1/engine/coordinator.py
index 44a346350fc81ac9b571fa515ae7f3cf2be04abe..8a01f06051349e50a4a3b1eff53b8f126fb14b86 100644
--- a/vllm/v1/engine/coordinator.py
+++ b/vllm/v1/engine/coordinator.py
@@ -246,9 +246,9 @@ class DPCoordinatorProc:
                         # Subscription message, on the other hand, is sent
                         # by each engine during initialization
                         publish_back.send(b"READY")
-                    else:
+                    elif buffer != b"\x00":
                         logger.error(
-                            "DP Coordinator receives unexpected message from engines"
+                            "DP Coordinator received unexpected message from engines"
                         )
 
                 if publish_front in events:
