diff --git a/vllm/v1/sample/ops/logprobs.py b/vllm/v1/sample/ops/logprobs.py
index cf36d46e13fdac4afab61000e476f79f4a6b712f..9cdcfa456a808831c2406e64da1ac23f96e26870 100644
--- a/vllm/v1/sample/ops/logprobs.py
+++ b/vllm/v1/sample/ops/logprobs.py
@@ -7,7 +7,7 @@ import torch
 from vllm.platforms import current_platform
 
 
-@torch.compile(dynamic=True, backend=current_platform.simple_compile_backend)
+@torch.compile(backend=current_platform.simple_compile_backend)
 def batched_count_greater_than(x: torch.Tensor, values: torch.Tensor) -> torch.Tensor:
     """
     Counts elements in each row of x that are greater than the corresponding
@@ -22,4 +22,6 @@ def batched_count_greater_than(x: torch.Tensor, values: torch.Tensor) -> torch.T
     Returns:
         torch.Tensor: A 1D tensor of shape (batch_size,) with the counts.
     """
+    torch._check(x.shape[0] >= 1)
+    torch._check(x.shape[0] == values.shape[0])
     return (x >= values).sum(-1)
